File: lib/services/geolocalizacion_service.dart
import 'package:dio/dio.dart';
import 'package:geolocator/geolocator.dart';
import '../models/geolocalizacion.dart';
import 'logger_service.dart';

/// Servicio de geolocalización
class GeolocalizacionService {
  static final GeolocalizacionService _instance = GeolocalizacionService._internal();
  factory GeolocalizacionService() => _instance;
  GeolocalizacionService._internal();

  final _logger = LoggerService();
  Dio? _dio;

  /// Inicializar servicio
  void initialize(Dio dio) {
    _dio = dio;
  }

  /// Obtener ubicación actual del dispositivo
  Future<PuntoGeo?> obtenerUbicacionActual() async {
    try {
      // Verificar permisos
      bool serviceEnabled = await Geolocator.isLocationServiceEnabled();
      if (!serviceEnabled) {
        _logger.warning('Servicio de ubicación deshabilitado');
        return null;
      }

      LocationPermission permission = await Geolocator.checkPermission();
      if (permission == LocationPermission.denied) {
        permission = await Geolocator.requestPermission();
        if (permission == LocationPermission.denied) {
          _logger.warning('Permiso de ubicación denegado');
          return null;
        }
      }

      if (permission == LocationPermission.deniedForever) {
        _logger.error('Permiso de ubicación denegado permanentemente');
        return null;
      }

      // Obtener posición
      final position = await Geolocator.getCurrentPosition(
        desiredAccuracy: LocationAccuracy.high,
      );

      _logger.info('Ubicación obtenida', data: {
        'lat': position.latitude,
        'lon': position.longitude,
      });

      return PuntoGeo(
        coordinates: [position.longitude, position.latitude],
      );
    } catch (e, stackTrace) {
      _logger.error('Error obteniendo ubicación', error: e, stackTrace: stackTrace);
      return null;
    }
  }

  /// Buscar entidades cercanas
  Future<List<EntidadCercana>> buscarCercanos({
    required double latitud,
    required double longitud,
    double radio = 10,
    String tipo = 'todos',
    int limit = 20,
  }) async {
    try {
      final response = await _dio!.get(
        '/geolocalizacion/cercanos',
        queryParameters: {
          'latitud': latitud,
          'longitud': longitud,
          'radio': radio,
          'tipo': tipo,
          'limit': limit,
        },
      );

      final entidades = (response.data['data'] as List)
          .map((e) => EntidadCercana.fromJson(e))
          .toList();

      _logger.info('Entidades cercanas encontradas', data: {'count': entidades.length});

      return entidades;
    } catch (e, stackTrace) {
      _logger.error('Error buscando entidades cercanas', error: e, stackTrace: stackTrace);
      rethrow;
    }
  }

  /// Calcular ruta entre dos puntos
  Future<RutaCalculada> calcularRuta({
    required PuntoGeo origen,
    required PuntoGeo destino,
    bool optimizar = true,
    bool evitarPeajes = false,
  }) async {
    try {
      final response = await _dio!.post(
        '/geolocalizacion/ruta',
        data: {
          'origen': origen.toJson(),
          'destino': destino.toJson(),
          'optimizar': optimizar,
          'evitarPeajes': evitarPeajes,
        },
      );

      final ruta = RutaCalculada.fromJson(response.data['data']);

      _logger.info('Ruta calculada', data: {
        'distancia': ruta.distancia,
        'duracion': ruta.duracion,
      });

      return ruta;
    } catch (e, stackTrace) {
      _logger.error('Error calculando ruta', error: e, stackTrace: stackTrace);
      rethrow;
    }
  }

  /// Calcular ruta múltiple optimizada
  Future<RutaMultipleCalculada> calcularRutaMultiple({
    required PuntoGeo origen,
    required List<PuntoGeo> destinos,
    bool optimizar = true,
    bool retornarAlOrigen = false,
  }) async {
    try {
      final response = await _dio!.post(
        '/geolocalizacion/ruta-multiple',
        data: {
          'origen': origen.toJson(),
          'destinos': destinos.map((d) => d.toJson()).toList(),
          'optimizar': optimizar,
          'retornarAlOrigen': retornarAlOrigen,
        },
      );

      final ruta = RutaMultipleCalculada.fromJson(response.data['data']);

      _logger.info('Ruta múltiple calculada', data: {
        'distanciaTotal': ruta.distanciaTotal,
        'duracionTotal': ruta.duracionTotal,
        'paradas': destinos.length,
      });

      return ruta;
    } catch (e, stackTrace) {
      _logger.error('Error calculando ruta múltiple', error: e, stackTrace: stackTrace);
      rethrow;
    }
  }

  /// Crear zona de cobertura
  Future<ZonaCobertura> crearZonaCobertura({
    required String nombre,
    String? descripcion,
    required String madrinaId,
    required String municipioId,
    required Poligono poligono,
    String? color,
    bool activo = true,
  }) async {
    try {
      final response = await _dio!.post(
        '/geolocalizacion/zonas',
        data: {
          'nombre': nombre,
          'descripcion': descripcion,
          'madrinaId': madrinaId,
          'municipioId': municipioId,
          'poligono': poligono.toJson(),
          'color': color,
          'activo': activo,
        },
      );

      final zona = ZonaCobertura.fromJson(response.data['data']);

      _logger.info('Zona de cobertura creada', data: {'id': zona.id});

      return zona;
    } catch (e, stackTrace) {
      _logger.error('Error creando zona de cobertura', error: e, stackTrace: stackTrace);
      rethrow;
    }
  }

  /// Obtener zonas de cobertura
  Future<List<ZonaCobertura>> obtenerZonasCobertura({String? municipioId}) async {
    try {
      final response = await _dio!.get(
        '/geolocalizacion/zonas',
        queryParameters: {
          if (municipioId != null) 'municipioId': municipioId,
        },
      );

      final zonas = (response.data['data'] as List)
          .map((z) => ZonaCobertura.fromJson(z))
          .toList();

      _logger.info('Zonas de cobertura obtenidas', data: {'count': zonas.length});

      return zonas;
    } catch (e, stackTrace) {
      _logger.error('Error obteniendo zonas de cobertura', error: e, stackTrace: stackTrace);
      rethrow;
    }
  }

  /// Obtener estadísticas de geolocalización
  Future<EstadisticasGeolocalizacion> obtenerEstadisticas() async {
    try {
      final response = await _dio!.get('/geolocalizacion/estadisticas');
      return EstadisticasGeolocalizacion.fromJson(response.data['data']);
    } catch (e, stackTrace) {
      _logger.error('Error obteniendo estadísticas', error: e, stackTrace: stackTrace);
      rethrow;
    }
  }

  /// Obtener heatmap
  Future<MapaHeatmap> obtenerHeatmap({
    required String tipo,
    String? municipioId,
    String? fechaInicio,
    String? fechaFin,
  }) async {
    try {
      final response = await _dio!.get(
        '/geolocalizacion/heatmap',
        queryParameters: {
          'tipo': tipo,
          if (municipioId != null) 'municipioId': municipioId,
          if (fechaInicio != null) 'fechaInicio': fechaInicio,
          if (fechaFin != null) 'fechaFin': fechaFin,
        },
      );

      return MapaHeatmap.fromJson(response.data['data']);
    } catch (e, stackTrace) {
      _logger.error('Error obteniendo heatmap', error: e, stackTrace: stackTrace);
      rethrow;
    }
  }

  /// Calcular distancia entre dos puntos (Haversine)
  double calcularDistancia(
    double lat1,
    double lon1,
    double lat2,
    double lon2,
  ) {
    return Geolocator.distanceBetween(lat1, lon1, lat2, lon2) / 1000; // km
  }

  /// Verificar si el servicio de ubicación está habilitado
  Future<bool> isLocationServiceEnabled() async {
    return await Geolocator.isLocationServiceEnabled();
  }

  /// Verificar permisos de ubicación
  Future<LocationPermission> checkPermission() async {
    return await Geolocator.checkPermission();
  }

  /// Solicitar permisos de ubicación
  Future<LocationPermission> requestPermission() async {
    return await Geolocator.requestPermission();
  }

  /// Abrir configuración de ubicación
  Future<bool> openLocationSettings() async {
    return await Geolocator.openLocationSettings();
  }

  /// Stream de ubicación en tiempo real
  Stream<Position> getPositionStream({
    LocationAccuracy accuracy = LocationAccuracy.high,
    int distanceFilter = 10, // metros
  }) {
    return Geolocator.getPositionStream(
      locationSettings: LocationSettings(
        accuracy: accuracy,
        distanceFilter: distanceFilter,
      ),
    );
  }
}

File: S/aplicacionWZC/madres-digitales-backend/src/services/geolocalizacion.service.ts
import prisma from '../config/database';
import {
  BuscarCercanosDTO,
  CalcularRutaDTO,
  CalcularRutaMultipleDTO,
  CrearZonaCoberturaDTO,
  ActualizarZonaCoberturaDTO,
  VerificarPuntoEnZonaDTO,
  ObtenerHeatmapDTO,
  ObtenerClustersDTO,
  AnalizarCoberturaDTO,
  EntidadCercana,
  RutaCalculada,
  RutaMultipleCalculada,
  ZonaCobertura,
  EstadisticasGeolocalizacion,
  MapaHeatmap,
  Cluster,
  AnalisisCobertura,
  PuntoGeoDTO,
} from '../core/application/dtos/geolocalizacion.dto';
import { logger } from '../config/logger';
import { NotFoundError } from '../core/domain/errors/not-found.error';

export class GeolocalizacionService {
  /**
   * Buscar entidades cercanas usando fórmula de Haversine
   */
  async buscarCercanos(dto: BuscarCercanosDTO): Promise<EntidadCercana[]> {
    try {
      const resultados: EntidadCercana[] = [];

      // Buscar gestantes cercanas
      if (dto.tipo === 'gestantes' || dto.tipo === 'todos') {
        const gestantes = await prisma.$queryRaw<any[]>`
          SELECT 
            id,
            nombre,
            ubicacion,
            direccion,
            telefono,
            (
              6371 * acos(
                cos(radians(${dto.latitud})) * 
                cos(radians((ubicacion->>'coordinates')::json->1::text::float)) * 
                cos(radians((ubicacion->>'coordinates')::json->0::text::float) - radians(${dto.longitud})) + 
                sin(radians(${dto.latitud})) * 
                sin(radians((ubicacion->>'coordinates')::json->1::text::float))
              )
            ) AS distancia
          FROM gestantes
          WHERE ubicacion IS NOT NULL
          HAVING distancia <= ${dto.radio}
          ORDER BY distancia
          LIMIT ${dto.limit}
        `;

        resultados.push(
          ...gestantes.map((g) => ({
            id: g.id,
            tipo: 'gestante' as const,
            nombre: g.nombre,
            ubicacion: g.ubicacion,
            distancia: parseFloat(g.distancia),
            direccion: g.direccion,
            telefono: g.telefono,
          }))
        );
      }

      // Buscar IPS cercanas
      if (dto.tipo === 'ips' || dto.tipo === 'todos') {
        const ips = await prisma.$queryRaw<any[]>`
          SELECT 
            id,
            nombre,
            ubicacion,
            direccion,
            telefono,
            (
              6371 * acos(
                cos(radians(${dto.latitud})) * 
                cos(radians((ubicacion->>'coordinates')::json->1::text::float)) * 
                cos(radians((ubicacion->>'coordinates')::json->0::text::float) - radians(${dto.longitud})) + 
                sin(radians(${dto.latitud})) * 
                sin(radians((ubicacion->>'coordinates')::json->1::text::float))
              )
            ) AS distancia
          FROM ips
          WHERE ubicacion IS NOT NULL AND activo = true
          HAVING distancia <= ${dto.radio}
          ORDER BY distancia
          LIMIT ${dto.limit}
        `;

        resultados.push(
          ...ips.map((i) => ({
            id: i.id,
            tipo: 'ips' as const,
            nombre: i.nombre,
            ubicacion: i.ubicacion,
            distancia: parseFloat(i.distancia),
            direccion: i.direccion,
            telefono: i.telefono,
          }))
        );
      }

      // Buscar madrinas cercanas
      if (dto.tipo === 'madrinas' || dto.tipo === 'todos') {
        const madrinas = await prisma.$queryRaw<any[]>`
          SELECT 
            u.id,
            u.nombre,
            u.ubicacion,
            u.telefono,
            (
              6371 * acos(
                cos(radians(${dto.latitud})) * 
                cos(radians((u.ubicacion->>'coordinates')::json->1::text::float)) * 
                cos(radians((u.ubicacion->>'coordinates')::json->0::text::float) - radians(${dto.longitud})) + 
                sin(radians(${dto.latitud})) * 
                sin(radians((u.ubicacion->>'coordinates')::json->1::text::float))
              )
            ) AS distancia
          FROM usuarios u
          WHERE u.ubicacion IS NOT NULL AND u.rol = 'madrina'
          HAVING distancia <= ${dto.radio}
          ORDER BY distancia
          LIMIT ${dto.limit}
        `;

        resultados.push(
          ...madrinas.map((m) => ({
            id: m.id,
            tipo: 'madrina' as const,
            nombre: m.nombre,
            ubicacion: m.ubicacion,
            distancia: parseFloat(m.distancia),
            telefono: m.telefono,
          }))
        );
      }

      // Ordenar por distancia y limitar
      resultados.sort((a, b) => a.distancia - b.distancia);
      return resultados.slice(0, dto.limit);
    } catch (error: any) {
      logger.error('Error buscando entidades cercanas', { error, dto });
      throw error;
    }
  }

  /**
   * Calcular ruta entre dos puntos (simplificado)
   */
  async calcularRuta(dto: CalcularRutaDTO): Promise<RutaCalculada> {
    try {
      // Calcular distancia usando Haversine
      const distancia = this._calcularDistanciaHaversine(
        dto.origen.coordinates[1],
        dto.origen.coordinates[0],
        dto.destino.coordinates[1],
        dto.destino.coordinates[0]
      );

      // Estimar duración (asumiendo 40 km/h promedio en zona rural)
      const duracion = (distancia / 40) * 60; // minutos

      // Generar puntos intermedios (simplificado - línea recta)
      const puntos: PuntoGeoDTO[] = [
        dto.origen,
        dto.destino,
      ];

      logger.info('Ruta calculada', { distancia, duracion });

      return {
        distancia,
        duracion,
        puntos,
        instrucciones: [
          `Dirigirse hacia el destino (${distancia.toFixed(2)} km)`,
          'Llegada al destino',
        ],
      };
    } catch (error: any) {
      logger.error('Error calculando ruta', { error, dto });
      throw error;
    }
  }

  /**
   * Calcular ruta múltiple optimizada
   */
  async calcularRutaMultiple(dto: CalcularRutaMultipleDTO): Promise<RutaMultipleCalculada> {
    try {
      let destinos = [...dto.destinos];
      let orden: number[] = [];
      let distanciaTotal = 0;
      let duracionTotal = 0;
      const rutas: RutaCalculada[] = [];

      if (dto.optimizar) {
        // Algoritmo del vecino más cercano
        let puntoActual = dto.origen;
        const destinosRestantes = destinos.map((d, i) => ({ destino: d, indice: i }));

        while (destinosRestantes.length > 0) {
          let menorDistancia = Infinity;
          let indiceMasCercano = 0;

          destinosRestantes.forEach((item, idx) => {
            const distancia = this._calcularDistanciaHaversine(
              puntoActual.coordinates[1],
              puntoActual.coordinates[0],
              item.destino.coordinates[1],
              item.destino.coordinates[0]
            );

            if (distancia < menorDistancia) {
              menorDistancia = distancia;
              indiceMasCercano = idx;
            }
          });

          const destinoMasCercano = destinosRestantes[indiceMasCercano];
          orden.push(destinoMasCercano.indice);

          const ruta = await this.calcularRuta({
            origen: puntoActual,
            destino: destinoMasCercano.destino,
            optimizar: true,
            evitarPeajes: dto.optimizar,
          });

          rutas.push(ruta);
          distanciaTotal += ruta.distancia;
          duracionTotal += ruta.duracion;

          puntoActual = destinoMasCercano.destino;
          destinosRestantes.splice(indiceMasCercano, 1);
        }

        // Si debe retornar al origen
        if (dto.retornarAlOrigen) {
          const rutaRetorno = await this.calcularRuta({
            origen: puntoActual,
            destino: dto.origen,
            optimizar: true,
            evitarPeajes: dto.optimizar,
          });

          rutas.push(rutaRetorno);
          distanciaTotal += rutaRetorno.distancia;
          duracionTotal += rutaRetorno.duracion;
        }
      } else {
        // Sin optimizar - orden original
        orden = destinos.map((_, i) => i);
        let puntoActual = dto.origen;

        for (const destino of destinos) {
          const ruta = await this.calcularRuta({
            origen: puntoActual,
            destino,
            optimizar: false,
            evitarPeajes: false,
          });

          rutas.push(ruta);
          distanciaTotal += ruta.distancia;
          duracionTotal += ruta.duracion;
          puntoActual = destino;
        }
      }

      logger.info('Ruta múltiple calculada', { distanciaTotal, duracionTotal, paradas: orden.length });

      return {
        distanciaTotal,
        duracionTotal,
        orden,
        rutas,
      };
    } catch (error: any) {
      logger.error('Error calculando ruta múltiple', { error, dto });
      throw error;
    }
  }

  /**
   * Crear zona de cobertura
   */
  async crearZonaCobertura(dto: CrearZonaCoberturaDTO): Promise<ZonaCobertura> {
    try {
      const zona = await prisma.zonaCobertura.create({
        data: {
          nombre: dto.nombre,
          descripcion: dto.descripcion,
          municipio_id: dto.madrinaId,
          municipio_id: dto.municipioId,
          poligono: dto.poligono,
          color: dto.color || this._generarColorAleatorio(),
          activo: dto.activo,
        },
      });

      logger.info('Zona de cobertura creada', { zonaId: zona.id });

      return this._mapZonaCobertura(zona);
    } catch (error: any) {
      logger.error('Error creando zona de cobertura', { error, dto });
      throw error;
    }
  }

  /**
   * Obtener zonas de cobertura
   */
  async obtenerZonasCobertura(municipioId?: string): Promise<ZonaCobertura[]> {
    try {
      const where: any = { activo: true };
      if (municipioId) {
        where.municipio_id = municipioId;
      }

      const zonas = await prisma.zonaCobertura.findMany({
        where,
        orderBy: { fecha_creacion: 'desc' },
      });

      return zonas.map(this._mapZonaCobertura);
    } catch (error: any) {
      logger.error('Error obteniendo zonas de cobertura', { error });
      throw error;
    }
  }

  /**
   * Calcular distancia usando fórmula de Haversine
   */
  private _calcularDistanciaHaversine(
    lat1: number,
    lon1: number,
    lat2: number,
    lon2: number
  ): number {
    const R = 6371; // Radio de la Tierra en km
    const dLat = this._toRad(lat2 - lat1);
    const dLon = this._toRad(lon2 - lon1);

    const a =
      Math.sin(dLat / 2) * Math.sin(dLat / 2) +
      Math.cos(this._toRad(lat1)) *
        Math.cos(this._toRad(lat2)) *
        Math.sin(dLon / 2) *
        Math.sin(dLon / 2);

    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    return R * c;
  }

  private _toRad(degrees: number): number {
    return degrees * (Math.PI / 180);
  }

  private _generarColorAleatorio(): string {
    const colores = ['#FF6B6B', '#4ECDC4', '#45B7D1', '#FFA07A', '#98D8C8', '#F7DC6F'];
    return colores[Math.floor(Math.random() * colores.length)];
  }

  private _mapZonaCobertura(zona: any): ZonaCobertura {
    return {
      id: zona.id,
      nombre: zona.nombre,
      descripcion: zona.descripcion,
      madrinaId: zona.municipio_id,
      municipioId: zona.municipio_id,
      poligono: zona.poligono,
      color: zona.color,
      activo: zona.activo,
      createdAt: zona.fecha_creacion,
      updatedAt: zona.updated_at,
    };
  }
}

